### PR TITLE
[Model] Pipeline parallel support for JAIS

### PR DESCRIPTION
**Pull request by mrbesher**
PP support for jais

Tested on "inceptionai/jais-family-590m" with pp=2 tp=1 in the following environment:
```
- Platform: Linux-6.5.0-21-generic-x86_64-with-glibc2.35
- Python version: 3.10.14
- Numpy version: 1.26.4
- PyTorch version (GPU?): 2.4.0+cu121 (True)
- System RAM: 251.53 GB
- GPU type: NVIDIA RTX A5000 (2 GPUs)
```

This is important as JAIS has models that do not support tensor parallelism at all (e.g.`inceptionai/jais-13b-chat` has `n_inner=13653` which is not divisible by 2).

Credit: Adapted changes from https://github.com/vllm-project/vllm/pull/6924

### CODE REVIEW

**Comment by youkaichao** [comment]
thanks for the contribution! please fix the format issue.

and cc @andoorve for review.

**Comment by mrbesher** [comment]
Thanks and sorry for the format issues.
@andoorve  can you review please?

**Comment by youkaichao** [comment]
I tested this locally by running `tests/distributed/test_pipeline_parallel.py` and replace the model with `inceptionai/jais-13b-chat` . It passed all the tests.

I will hand it over to @andoorve for the final pass.

File: vllm/model_executor/models/jais.py [hunk 1]
```diff
@@ -243,19 +251,33 @@ def forward(
         position_ids: torch.Tensor,
         kv_caches: List[torch.Tensor],
         attn_metadata: AttentionMetadata,
+        intermediate_tensors: Optional[IntermediateTensors] = None,
+        inputs_embeds: Optional[torch.Tensor] = None,
```
Comment (andoorve) [inline vllm/model_executor/models/jais.py:255]: We don't need input_embeds here. Simply calculate it the same way we have previously in the first PP rank

File: vllm/model_executor/models/jais.py [hunk 2]
```diff
@@ -243,19 +251,33 @@ def forward(
...
     ) -> torch.Tensor:
```
Comment (andoorve) [inline vllm/model_executor/models/jais.py:256]: Type annotation needed, union with IntermediateTensors

File: vllm/model_executor/models/jais.py [hunk 3]
```diff
@@ -292,7 +314,7 @@ def forward(
         intermediate_tensors: Optional[IntermediateTensors] = None,
     ) -> torch.Tensor:
```
Comment (andoorve) [inline vllm/model_executor/models/jais.py:315]: Type annotation needed, union with IntermediateTensors

File: vllm/model_executor/models/jais.py [hunk 4]
```diff
@@ -243,19 +251,33 @@ def forward(
...
         attn_metadata: AttentionMetadata,
+        intermediate_tensors: Optional[IntermediateTensors] = None,
+        inputs_embeds: Optional[torch.Tensor] = None,
     ) -> torch.Tensor:
-        inputs_embeds = self.wte(input_ids)
-        if self.wpe is not None:
-            position_embeds = self.wpe(position_ids)
-            hidden_states = inputs_embeds + position_embeds
+        if get_pp_group().is_first_rank:
+            if inputs_embeds is not None:
```
Comment (andoorve) [inline vllm/model_executor/models/jais.py:258]: With reference to the above comment about input_embeds, we don't need this conditional here.

**Comment by andoorve** [comment]
Added minor comments, please take a look

**Review by andoorve — APPROVED** [review]
LGTM!

**Comment by youkaichao** [comment]
Thanks for the contribution!

**Pull request merged**